pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/security/HardCodedCryptoKeyRule.java

```diff
@@ -0,0 +1,161 @@
+import net.sourceforge.pmd.lang.java.symboltable.VariableNameDeclaration;
+import net.sourceforge.pmd.lang.symboltable.NameOccurrence;
+
+/**
+ * Finds hard coded encryption keys that are passed to
+ * javax.crypto.spec.SecretKeySpec(key, algorithm).
+ * 
+ * @author sergeygorbaty
+ *
+ */
+public class HardCodedCryptoKeyRule extends AbstractJavaRule {
+
+    public HardCodedCryptoKeyRule() {
+        addRuleChainVisit(ASTClassOrInterfaceBodyDeclaration.class);
+    }
+
+    @Override
+    public Object visit(ASTClassOrInterfaceBodyDeclaration node, Object data) {
+        Set<ASTLocalVariableDeclaration> foundLocalVars = new HashSet<>();
+
+        // find new javax.crypto.spec.SecretKeySpec("literal".getBytes(),...);
+        List<ASTAllocationExpression> allocations = node.findDescendantsOfType(ASTAllocationExpression.class);
+        for (ASTAllocationExpression allocation : allocations) {
+
+            ASTClassOrInterfaceType declClassName = allocation.getFirstDescendantOfType(ASTClassOrInterfaceType.class);
+            if (declClassName != null) {
+                Class<?> foundClass = declClassName.getType();
+                if (foundClass != null && javax.crypto.spec.SecretKeySpec.class.isAssignableFrom(foundClass)) {
+                    ASTPrimaryExpression init = allocation.getFirstDescendantOfType(ASTPrimaryExpression.class);
+                    if (init != null) {
+                        ASTPrimaryPrefix prefix = init.getFirstChildOfType(ASTPrimaryPrefix.class);
+                        if (prefix != null) {
+                            ASTLiteral literal = prefix.getFirstChildOfType(ASTLiteral.class);
+                            if (literal != null) {
+                                addViolation(data, literal);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // variables that were passed to SecretKeySpec
+        Set<String> passedInIvVarNames = Util.findVariablesPassedToAnyParam(node,
+                javax.crypto.spec.SecretKeySpec.class);
+
+        List<ASTLocalVariableDeclaration> localVars = node.findDescendantsOfType(ASTLocalVariableDeclaration.class);
+        for (ASTLocalVariableDeclaration localVar : localVars) {
+            foundLocalVars.addAll(extractPrimitiveTypes(localVar));
+        }
+
+        Map<VariableNameDeclaration, List<NameOccurrence>> globalDecls = node.getScope()
+                .getDeclarations(VariableNameDeclaration.class);
+
+        for (VariableNameDeclaration fieldVar : globalDecls.keySet()) {
+            if (passedInIvVarNames.contains(fieldVar.getNode().getImage())) {
+                ASTVariableDeclarator var = fieldVar.getNode().getFirstParentOfType(ASTVariableDeclarator.class);
+                if (var != null) {
+                    validateProperKey(data, var.getFirstDescendantOfType(ASTVariableInitializer.class));
+                }
+            }
+        }
```

Comment:
these 3 for loops are somewhat redundant and fail to deal with nested scopes properly (specially in the presence of lambdas and nested classes).

I'd just keep this second loop, but making sure to also iterate over the parent scopes.

If reworking this to start directly from the allocation instead of `ASTClassOrInterfaceBodyDeclaration` you automatically start from the inner-most scope and can easily work your way up